File: haskellapp1/Game.hs
-- 推箱子核心逻辑（Haskell 教学）
module Game
  ( GameState(..)
  , fromRows
  , tryMove
  , undo
  , renderAscii
  ) where

import Data.List (foldl')
import Data.Set (Set)
import qualified Data.Set as Set

type Pos = (Int, Int)
type Cell = String

data Hist = Hist
  { hPlayer :: Pos
  , hBoxFrom :: Maybe Cell
  , hBoxTo :: Maybe Cell
  } deriving (Show)

data GameState = GameState
  { walls :: Set Cell
  , goals :: Set Cell
  , boxes :: Set Cell
  , player :: Pos
  , moves :: Int
  , won :: Bool
  , width :: Int
  , height :: Int
  , levelIndex :: Int
  , hist :: [Hist]
  } deriving (Show)

key :: Int -> Int -> Cell
key x y = show x ++ "," ++ show y

fromRows :: [String] -> Int -> GameState
fromRows rows idx =
  let cells = [ (x, y, ch)
              | (y, row) <- zip [0..] rows
              , (x, ch) <- zip [0..] row
              ]
      maxX = maximum (0 : [x | (x, _, _) <- cells])
      maxY = maximum (0 : [y | (_, y, _) <- cells])
      go (ws, gs, bs, pl) (x, y, ch) =
        let k = key x y
        in case ch of
             '#' -> (Set.insert k ws, gs, bs, pl)
             '.' -> (ws, Set.insert k gs, bs, pl)
             '$' -> (ws, gs, Set.insert k bs, pl)
             '*' -> (ws, Set.insert k gs, Set.insert k bs, pl)
             '@' -> (ws, gs, bs, (x, y))
             '+' -> (ws, Set.insert k gs, bs, (x, y))
             _   -> (ws, gs, bs, pl)
      (ws, gs, bs, pl) = foldl' go (Set.empty, Set.empty, Set.empty, (0, 0)) cells
  in GameState ws gs bs pl 0 False (maxX + 1) (maxY + 1) idx []

checkWin :: Set Cell -> Set Cell -> Bool
checkWin bs gs = all (`Set.member` gs) (Set.toList bs)

tryMove :: Int -> Int -> GameState -> GameState
tryMove dx dy s
  | won s = s
  | Set.member nk (walls s) = s
  | Set.member nk (boxes s) =
      let bx = nx + dx
          by = ny + dy
          bk = key bx by
      in if Set.member bk (walls s) || Set.member bk (boxes s)
           then s
           else
             let boxes' = Set.insert bk (Set.delete nk (boxes s))
                 moves' = moves s + 1
             in s
               { boxes = boxes'
               , player = (nx, ny)
               , moves = moves'
               , won = checkWin boxes' (goals s)
               , hist = Hist (player s) (Just nk) (Just bk) : hist s
               }
  | otherwise =
      s
        { player = (nx, ny)
        , hist = Hist (player s) Nothing Nothing : hist s
        }
  where
    (px, py) = player s
    nx = px + dx
    ny = py + dy
    nk = key nx ny

undo :: GameState -> GameState
undo s
  | won s || null (hist s) = s
  | otherwise = go (hist s) (player s)
  where
    go [] pl = s { player = pl, hist = [] }
    go (h:rest) pl =
      case hBoxFrom h of
        Nothing -> go rest (hPlayer h)
        Just bf ->
          let bt = maybe bf id (hBoxTo h)
              boxes' = Set.insert bf (Set.delete bt (boxes s))
          in s
            { player = hPlayer h
            , boxes = boxes'
            , moves = max 0 (moves s - 1)
            , won = False
            , hist = rest
            }

renderAscii :: GameState -> String
renderAscii s = unlines
  [ [ cell x y | x <- [0 .. width s - 1] ]
  | y <- [0 .. height s - 1]
  ]
  where
    cell x y
      | player s == (x, y) = if Set.member k (goals s) then '+' else '@'
      | Set.member k (boxes s) = if Set.member k (goals s) then '*' else '$'
      | Set.member k (walls s) = '#'
      | Set.member k (goals s) = '.'
      | otherwise = ' '
      where
        k = key x y

File: elixirapp1/game.exs
# 推箱子核心逻辑（Elixir 教学）

defmodule Sokoban.Game do
  defstruct walls: MapSet.new(), goals: MapSet.new(), boxes: MapSet.new(),
            px: 0, py: 0, moves: 0, won: false, width: 0, height: 0, hist: []

  def key(x, y), do: "#{x},#{y}"

  def from_rows(rows, _index \\ 0) do
    {walls, goals, boxes, px, py, max_x, max_y} =
      rows
      |> Enum.with_index()
      |> Enum.reduce({MapSet.new(), MapSet.new(), MapSet.new(), 0, 0, 0, 0}, fn {row, y}, acc ->
        row
        |> String.graphemes()
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {ch, x}, {w, g, b, px, py, mx, my} ->
          mx = max(mx, x)
          my = max(my, y)
          k = key(x, y)
          case ch do
            "#" -> {MapSet.put(w, k), g, b, px, py, mx, my}
            "." -> {w, MapSet.put(g, k), b, px, py, mx, my}
            "$" -> {w, g, MapSet.put(b, k), px, py, mx, my}
            "*" -> {w, MapSet.put(g, k), MapSet.put(b, k), px, py, mx, my}
            "@" -> {w, g, b, x, y, mx, my}
            "+" -> {w, MapSet.put(g, k), b, x, y, mx, my}
            _ -> {w, g, b, px, py, mx, my}
          end
        end)
      end)

    %__MODULE__{
      walls: walls, goals: goals, boxes: boxes,
      px: px, py: py, width: max_x + 1, height: max_y + 1
    }
  end

  defp check_win(boxes, goals), do: Enum.all?(boxes, &MapSet.member?(goals, &1))

  def try_move(%{won: true} = s, _, _), do: s

  def try_move(s, dx, dy) do
    nx = s.px + dx
    ny = s.py + dy
    nk = key(nx, ny)

    cond do
      MapSet.member?(s.walls, nk) ->
        s

      MapSet.member?(s.boxes, nk) ->
        bx = nx + dx
        by = ny + dy
        bk = key(bx, by)

        if MapSet.member?(s.walls, bk) or MapSet.member?(s.boxes, bk) do
          s
        else
          boxes = s.boxes |> MapSet.delete(nk) |> MapSet.put(bk)

          %{
            s
            | boxes: boxes,
              px: nx,
              py: ny,
              moves: s.moves + 1,
              won: check_win(boxes, s.goals),
              hist: [%{px: s.px, py: s.py, bf: nk, bt: bk} | s.hist]
          }
        end

      true ->
        %{s | px: nx, py: ny, hist: [%{px: s.px, py: s.py, bf: nil, bt: nil} | s.hist]}
    end
  end

  def undo(%{won: true} = s), do: s
  def undo(%{hist: []} = s), do: s

  def undo(s), do: do_undo(s.hist, s)

  defp do_undo([], s), do: %{s | hist: [], won: false}

  defp do_undo([%{bf: nil, px: px, py: py} | rest], s),
    do: do_undo(rest, %{s | px: px, py: py, hist: rest})

  defp do_undo([%{bf: bf, bt: bt, px: px, py: py} | rest], s) when not is_nil(bf) do
    boxes = s.boxes |> MapSet.delete(bt) |> MapSet.put(bf)

    %{
      s
      | boxes: boxes,
        px: px,
        py: py,
        moves: max(0, s.moves - 1),
        won: false,
        hist: rest
    }
  end

  def render_ascii(s) do
    for y <- 0..(s.height - 1), into: "" do
      row =
        for x <- 0..(s.width - 1), into: "" do
          k = key(x, y)

          cond do
            s.px == x and s.py == y -> if(MapSet.member?(s.goals, k), do: "+", else: "@")
            MapSet.member?(s.boxes, k) -> if(MapSet.member?(s.goals, k), do: "*", else: "$")
            MapSet.member?(s.walls, k) -> "#"
            MapSet.member?(s.goals, k) -> "."
            true -> " "
          end
        end

      row <> "\n"
    end
  end
end
